pwndbg/commands/mmap.py

```diff
@@ -0,0 +1,241 @@
+)
+parser.add_argument(
+    "--force", "-f", help="Force potentially unsafe actions to happen", action="store_true"
+)
+
+
+prot_dict = {
+    "PROT_NONE": 0x0,
+    "PROT_READ": 0x1,
+    "PROT_WRITE": 0x2,
+    "PROT_EXEC": 0x4,
+}
+
+flag_dict = {
+    "MAP_SHARED": 0x1,
+    "MAP_PRIVATE": 0x2,
+    "MAP_SHARED_VALIDATE": 0x3,
+    "MAP_FIXED": 0x10,
+    "MAP_ANONYMOUS": 0x20,
+}
+
+
+def prot_str_to_val(protstr):
+    """Heuristic to convert PROT_EXEC|PROT_WRITE to integer value."""
+    prot_int = 0
+    for k, v in prot_dict.items():
+        if k in protstr:
+            prot_int |= v
+    return prot_int
+
+
+def flag_str_to_val(flagstr):
+    """Heuristic to convert MAP_SHARED|MAP_FIXED to integer value."""
+    flag_int = 0
+    for k, v in flag_dict.items():
+        if k in flagstr:
+            flag_int |= v
+    return flag_int
+
+
+def parse_str_or_int(val, parser):
+    """
+    Try parsing a string with one of the parsers above or by converting it to
+    an int, or passes the value through if it is already an integer.
+    """
+    if type(val) is str:
+        candidate = parser(val)
+        if candidate != 0:
+            return candidate
+        return int(val, 0)
+    elif type(val) is int:
+        return val
+    else:
+        # Getting here is a bug, we shouldn't be seeing other types at all.
+        raise TypeError(f"invalid type for value: {type(val)}")
+
+
+@pwndbg.commands.ArgparsedCommand(parser, category=CommandCategory.MEMORY)
+@pwndbg.commands.OnlyWhenRunning
+def mmap(addr, length, prot=7, flags=0x22, fd=-1, offset=0, quiet=False, force=False) -> None:
+    try:
+        prot_int = parse_str_or_int(prot, prot_str_to_val)
+    except ValueError as e:
+        print(message.error(f'Invalid protection value "{prot}": {e}'))
+        return        
+
+    try:
+        flag_int = parse_str_or_int(flags, flag_str_to_val)
+    except ValueError as e:
+        print(message.error(f'Invalid flags value "{flags}": {e}'))
+        return
+
+    aligned_addr = int(pwndbg.lib.memory.page_align(addr))
+    if flag_int & flag_dict["MAP_FIXED"] != 0:
+        # When using MAP_FIXED, it's only safe to call mmap(2) when the address
+        # overlaps no other maps. We want to make sure that, unless the user
+        # _really_ knows what they're doing, this call will be safe.
+        #
+        # Additionally, it's nice to highlight cases where the call is likely
+        # to fail because the address is not properly aligned.
+        addr = int(addr)
+        if addr != aligned_addr and not quiet:
+            print(
+                message.warn(
+                    f"""\
+Address {addr:#x} is not properly aligned. Calling mmap with MAP_FIXED and an
+unaligned address is likely to fail. Consider using the address {aligned_addr:#x}
+instead.\
+"""
+                )
+            )
+
+        page = pwndbg.lib.memory.Page(addr, int(length), 0, 0)
+        collisions = []
```

Comment:
In theory, there should never be more than a single collision (in practice if we have broken vmmap info there will be). We can safely assume a single collision here. Its fine if we print a single one if there are more

Reply:
I don't think I follow. If any range is permissible for our mmap, it could, in the worst case, technically collide with all of the mappings, no? Nothing's really stopping you from trying to call this with `MAP_FIXED` for `[0, 0xffffffffffffffff[` (I know in practice there are limitations on covering the whole range, but this same logic applies for smaller ranges too).

Reply:
Oh... you are right. For some reason I was thinking that we map a single address instead of a range. Forgive my stupidness :)